Enchance fix_includes.py to properly scan allman namespaces

chri7325 opened the pull request:
The fix_includes.py script has been tailored to work for
namespaces that have { on the same line as the namespace.
For styles that do not follow this format, namespace
detection does not work properly as the line numbers and
regex statement are not able to be evaluated.  Logic has
been added to detect when an allman namespace has been
found and shifts the line numbers appropriately.

Signed-off-by: Christian Venegas <cvenegas@esri.com>

Fixes issue https://github.com/include-what-you-use/include-what-you-use/issues/557

kimgr commented on fix_includes.py:
```diff
@@ -968,24 +979,37 @@ class Myclass;
+    elif line_info.type == _NAMESPACE_START_ALLMAN_RE:
+      # For Allman namespaces, keep the start line and increment
+      # the namespace depths when the actual brace is encountered.
+      namespace_depth = 0
```
It's a little disconcerting that `namespace_depth` is unbound if type is neither `_NAMESPACE_START_RE` nor `_NAMESPACE_START_ALLMAN_RE`. After some scanning, I found that that's guaranteed by the conditional above, but maybe it would be nice to add an `else` leg with an `assert False` to make sure that invariant doesn't break.

chri7325 replied:
@kimgr great point.  I've added an assert to guard against future additions.

kimgr replied:
:+1:

kimgr commented on fix_includes_test.py:
```diff
@@ -335,6 +335,80 @@ class Alert;                         ///-
+    self.RegisterFileContents({'empty_internal_namespace': infile})
+    self.ProcessAndTest(iwyu_output)
+
+  def testRemoveEmptyAllmanNamespace(self):
```
I haven't quite understood what the "continued Allman" regex matches -- should there be a test with nested namespaces to cover that code? E.g.
```
namespace A { namespace B { namespace C
{
}
}
}
```

or however it looks.

chri7325 replied:
@kimgr the continued allman regex simply matches the { on the next line.
namespace A // allman start re
{ // allman continued re
} // namespace end re

It is only triggered when an allman namespace is found (namespace without any { on the same line.  I'll add that test case and see if I can handle it somehow but it is a mix of both styles so it might be a bit tricky.

chri7325 replied:
Sorry, just to clarify a bit more, the "allman" style is something I'm getting from clang-format.  We use allman style for our clang-format option which states to always break before braces so for nested namespaces, it would be like the test cases supplied here
namespace A
{
namespace B
{
namespace C
{
} // namespace C
} // namespace B
} // namespace A

In order to support this formatting style, I added the regex statements for allman style as well as a regex for finding the appropriate continuation { line.

chri7325 replied:
Logic and test cases were added for mixed namespaces.  Now all namespace styles should be supported (hopefully haha).

kimgr commented on fix_includes.py:
```diff
@@ -1344,6 +1389,19 @@ def _GetFirstNamespaceLevelReorderSpan(file_lines):
 
+    elif line_info.type in (_NAMESPACE_START_ALLMAN_RE,
+                            _NAMESPACE_START_MIXED_RE):
+      # Trim any possibly comments on namespace to properly build
```
Typo: possibly -> possible

chri7325 replied:
Woops, nice catch.

kimgr commented:
Cool, we're nearly there!

chri7325 commented:
@kimgr I've fixed the typo and rebased onto my last commit to keep things clean.  As for the line indicators, the way IWYU produces the output data is only to show which forward declares need to be removed.  They only function on the forward declare itself.  They only use multi-line indicators when you have forward declares that span multiple lines such as templated forward declares.  The actual removal of empty namespaces is done by the python script itself, not by the IWYU program which don't operate on them.  I have a bit more work that I was hoping to submit soon that does better empty namespace detection as well as multi-level namespace detection.  I've been running my enhancements over our entire codebase (~2 million loc) and haven't found any issues as of yet with the next PR request I was hoping to submit.  Please let me know if this would be okay and if I should open an issue first before submitting another PR.

kimgr commented:
No need to open a bug issue for the subsequent changes, just post a PR, please!

chri7325 commented:
@kimgr please be as picky as possible.  IWYU was an amazing tool that allowed us to reduce our build times in an nearly automated fashion so anything to keep the projects code quality high is a much appreciated.  I'll amend the messages right away.

kimgr commented on fix_includes.py:
```diff
@@ -706,6 +717,10 @@ def _CalculateLineTypesAndKeys(file_lines, iwyu_record):
     elif in_c_style_comment:
       line_info.type = _COMMENT_LINE_RE
+    elif (in_allman_or_mixed_namespace and
+         _NAMESPACE_CONTINUE_ALLMAN_MIXED_RE.match(line_info.line)):
```
Ouch, line is under-indented by one space!

kimgr commented on fix_includes.py:
```diff
@@ -968,24 +987,43 @@ class Myclass;
     line_info = file_lines[start_line]
-    if line_info.deleted or line_info.type != _NAMESPACE_START_RE:
+    if (line_info.deleted or 
+       (line_info.type != _NAMESPACE_START_RE and
```
Under-indented

chri7325 commented:
Updated

kimgr commented on fix_includes.py:
```diff
@@ -1189,7 +1227,9 @@ def _ShouldInsertBlankLine(decorated_move_span, next_decorated_move_span,
       next_line += 1
     return (next_line and next_line < len(file_lines) and
-            file_lines[next_line].type in (_NAMESPACE_START_RE, None))
+            file_lines[next_line].type in 
```
I'd align this as:
```
file_lines[next_line].type in (_NAMESPACE_START_RE,
                               _NAMESPACE_START_ALLMAN_RE,
                               _NAMESPACE_START_MIXED_RE,
                               None))
```

Not sure Python agrees that's a good thing.

kimgr commented on fix_includes_test.py:
```diff
@@ -299,6 +299,78 @@ class Baz;       ///-
+    self.ProcessAndTest(iwyu_output)
+
+  def testRemoveEmptyAllmanNamespace(self):
+    """Tests we remove a namespace with allman braces if we remove all fwd-decls inside it."""
```
Allman

kimgr commented on fix_includes_test.py:
```diff
@@ -335,6 +407,78 @@ class Alert;                         ///-
+    self.ProcessAndTest(iwyu_output)
+
+  def testRemovePartOfEmptyAllmanNamespace(self):
+    """Tests we remove a namespace with allman braces if empty, but not enclosing namespaces."""
```
Allman

kimgr commented on fix_includes_test.py:
```diff
@@ -2557,6 +2701,59 @@ class Bar;       ///-
+    self.ProcessAndTest(iwyu_output)
+
+  def testIterativeAllmanNamespaceDelete(self):
+    """Tests deleting an allman namespace with an emptied #ifdef inside it."""
```
Allman

chri7325 commented:
Updated.  I'm thinking if everything is okay now, I can squash everything into a single commit but I'm leaving them as is now so you can see what I just changed instead of rebasing now.  If you feel this is good, please also feel free to squash when merging too.

kimgr commented on fix_includes.py:
```diff
@@ -968,24 +987,43 @@ class Myclass;
-    if line_info.deleted or line_info.type != _NAMESPACE_START_RE:
+    if (line_info.deleted or 
+        (line_info.type != _NAMESPACE_START_RE and
+        line_info.type != _NAMESPACE_START_ALLMAN_RE and
```
These two lines should be indented one more space to align with the `line_info` above

kimgr commented:
Squashed and merged. Thank you for your contribution!
